### ISSUE TITLE
DeclareStrictTypesFixer on interfaces is superfluous

### ISSUE THREAD

**Issue by ghost** [PHP-CS-Fixer/PHP-CS-Fixer#8097]
### Feature request

DeclareStrictTypesFixer on interfaces does not make any sense and only makes noise and redundant diff.

### Contribution Checks

- [ ] I have verified if this feature request was already discussed
- [ ] I am familiar with ["Feature or bug?"](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/feature-or-bug.rst)

**Comment by SpacePossum** [comment]
would be a nice to have configuration to opt-in/opt-out for interfaces :+1:

**Comment by Wirone** [comment]
It was already mentioned [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/2062#issuecomment-1399281180), but was never added to the backlog. I think it's a good idea to make it possible to skip interfaces.

**Comment by MrYamous** [comment]
> Hi, this may be a trivial question but for the record: the fix is not as simple as `!$tokens->isTokenKindFound(T_INTERFACE)` inside `isCandidate()` because a file may contain other code than only an interface declaration, right?
> 
> If the above is "yes" then how one should approach solving this issue?
> 
> Thank you.

Hi, I thought the same thing the last time I looked at this issue

**Comment by SpacePossum** [comment]
>If the above is "yes" then how one should approach solving this issue?

Not so simple, but also not very hard. Something like;
- add on opt-in option so by default the rule acts as it acts now (any configurable rule can be used as example), to keep the BC promise
- don't change the `is candidate` method, it is not needed, in the `applyFix` method we need to change things

If the rule is configured to not insert declare strict for files containing _only_ interface declarations,
than we choose a different path in the `applyFix` method:

Determine if the file only contains interface declarations:
-  skip the PHP open tag (+ non meaningfuls)
-  skip imports (`use` etc.,) (+ non meaningfuls)
Now:
-  if next is T_INTERFACE, go to next `{`, proceed to block end `}`, repeat this until:
-  next token is non T_INTERFACE yet a meaningful one (for example T_CLASS, than the file contains both interface and class):
    - than _do insert_ the declare strict
- else if; end of tokens; than the file only contains an interface so do _not insert_ the declare strict

**Comment by keradus** [comment]
- I want to have each file having same beginning, eg strict type or license header. I do not want to have "style" depending on content. It's not unified and very confusing.

- if we have strong community voice to have the option to opt-out and not declare strict for interfaces, OK to introduce fixer configurable option for it - but want to hear that strong community voice first, not single Hero only

**Comment by kubawerlos** [comment]
> DeclareStrictTypesFixer on interfaces does not make any sense and only makes noise and redundant diff.

By noise do you mean the 24 characters of `declare(strict_types=1);`?

> I want to have each file having same beginning, eg strict type or license header. I do not want to have "style" depending on content. It's not unified and very confusing.

I fully agree with @keradus on this.

If the reasoning is to not add `declare(strict_types=1);` when it does not affect the behaviour, this class would also not need it:
```php
<?php
class Foo
{
    public const CRAZY_CONST = [
        'is this constant crazy?' => true,
        'how crazy on a scale of 1 to 10?' => 42,
        'what about nested arrays?' => ['sure' => ['nesting' => ['nesting' => ['nesting' => ['nesting' => ['nesting' => ['nesting' => 'ok, enought']]]]]]],
        -NAN => 'whyyy?',
        'maybe some calculations?' => 1 + 2 + 3,
    ];

    public static function bar(int $int, float $float): float
    {
        return $int + $float;
    }

    public static function baz(array $array): array
    {
        return array_map(
            static fn (int $x): int => 2 * $x,
            $array
        );
    }
}
```

**Comment by kubawerlos** [comment]
> Every thinking developer would ask himself "what this line is doing here if the interface will never call or be called any method.".

@javaDeveloperKid I "love" these kinds of generalizations. Did you ask "every thinking developer" about this or is it just your personal opinion and you think everyone has to agree with it?

Have you consider that some thinking developers my value consistency over inconsistency?

**Comment by SpacePossum** [comment]
>If the reasoning is to not add declare(strict_types=1); when it does not affect the behaviour, this class would also not need it:

this was proposed by Julian and heiglandreas but got rejected https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6688 

there is also https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/2062#issuecomment-1399281180

```
The reason why I brought up that citation though was because the declare_strict is only relevant within files that actually call other functions. As an interface does - by definition - only define function-signatures and never calls other functions. A declare(stict_types) is therefore always unnecessary in an interface-file.
```

which was :+1:  by Wirone

I would like to remove (and not add declare stricts) from files with only single interface definition

So not sure if `single Hero only`. I have no opinion about people who want or not wanted this as I see it as a preference.

**Comment by kubawerlos** [comment]
> I see it as a preference.

Exactly, there is no need to offend people who prefer to keep declare strict in files containing only interface definition.

**Comment by kubawerlos** [comment]
> Another developer comes by and asks himsef "why this developer added this line? Was there any reason?".

I wouldn't ask myself that, I'd assume the obvious - consistency.

> the problem of having strict types inside a file that is and will be always superfluous.

Like the code from my comment: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/8097#issuecomment-2323270660 - would you expect/want anyone to analyse if strict types there are needed or not?

> You mean "every thinking developer" as offendable? By thinking in this case I meant a person that does not ignore this line.

I bet some can find it offensive. It implies that other, unthinking, developers would not ask the question.

I'd like to ignore the line `<?php declare(strict_types=1);` (or 3 lines, if coding standards say so) and make a habit of treating all the code as strict one.

**Comment by keradus** [comment]
Thanks for your voice everyone.
I believe we understand one another claim and need and one won't convince the other more, anymore.

I would love to wait and hear other developers now, if any.
If no further interest by other community members, i will be into declining feature request (but hey, custom fixers are way to go exactly for such cases)
